Add size related functions (#499)
timescale/docs.timescale.com-content: api.md

> - [get_telemetry_report](#get_telemetry_report)
> - [histogram](#histogram)
> - [hypertable_compression_stats](#hypertable_compression_stats)

> > - [hypertable_detailed_size](#hypertable_size)
> > - [hypertable_size](#hypertable_size)
> > - [hypertable_index_size](#hypertable_index_size)

Reviewer: ```suggestion
> - [hypertable_detailed_size](#hypertable_detailed_size)
> - [hypertable_index_size](#hypertable_index_size)
> - [hypertable_size](#hypertable_size)
```

```
---

> ## approximate_row_count() :community_function:  [](approximate_row_count)

Reviewer: Isn't this in Apache-2 codebase, not TSL?  If so, it shoudn't be marked :community_function:

Author: yes, this is Apache. Removed.

information as a separate row per node. 

#### Required Arguments [](hypertable_detailed_size-required-arguments)

|Name|Description|
|---|---|
| `main_table` | (REGCLASS) Name of the hypertable |

#### Returns [](hypertable_detailed_size-returns)
|Column|Description|
|---|---|
|table_bytes|(BIGINT) Disk space used by main_table (like pg_relation_size(main_table))|
|index_bytes|(BIGINT) Disk space used by indexes|
|toast_bytes|(BIGINT) Disk space of toast tables|
|total_bytes|(BIGINT) Total disk space used by the specified table, including all indexes and TOAST data|
|node_name| (NAME) Node for which size is reported, applicable only to distributed hypertables|

#### Sample Usage [](hypertable_detailed_size-examples)
Get size information for a hypertable.
```sql
-- disttable is a distributed hypertable --
SELECT * FROM hypertable_detailed_size('disttable') ORDER BY node_name;

 table_bytes | index_bytes | toast_bytes | total_bytes |  node_name
-------------+-------------+-------------+-------------+-------------
       16384 |       32768 |           0 |       49152 | data_node_1
        8192 |       16384 |           0 |       24576 | data_node_2

```
---

> ## chunks_detailed_size()  :community_function: [](chunks_detailed_size)

Reviewer: Remove :community_function:

|---|---|
|chunk_schema| (NAME) Schema name of the chunk |
|chunk_name| (NAME) Name of the chunk|
|table_bytes|(BIGINT) Disk space used by the chunk table|
|index_bytes|(BIGINT) Disk space used by indexes|
|toast_bytes|(BIGINT) Disk space of toast tables|
|total_bytes|(BIGINT) Total disk space used by the chunk, including all indexes and TOAST data|
|node_name| (NAME) Node for which size is reported, applicable only to distributed hypertables|

#### Sample Usage [](chunks_detailed_size-examples)
```sql
SELECT * FROM chunks_detailed_size('dist_table')
  ORDER BY chunk_name, node_name;

     chunk_schema      |      chunk_name       | table_bytes | index_bytes | toast_bytes | total_bytes |       node_name
-----------------------+-----------------------+-------------+-------------+-------------+-------------+-----------------------
 _timescaledb_internal | _dist_hyper_1_1_chunk |        8192 |       32768 |           0 |       40960 | db_node1
 _timescaledb_internal | _dist_hyper_1_2_chunk |        8192 |       32768 |           0 |       40960 | db_node2
 _timescaledb_internal | _dist_hyper_1_3_chunk |        8192 |       32768 |           0 |       40960 | db_node3
```
---

> ## hypertable_size() :community_function: [](hypertable_size)

Reviewer: Remove :community_function:

Get total size of hypertable i.e. the sum of the size for the table itself, 
any indexes on the table, and any toast tables. The size is reported in bytes. 
This is equivalent to computing the sum of `total_bytes` column from the 
output of `hypertable_detailed_size` function.

#### Required Arguments [](hypertable_size-required-arguments)

|Name|Description|
|---|---|
| `main_table` | (REGCLASS) Name of the hypertable |

#### Returns [](hypertable_size-returns)
(BIGINT) Total disk space used by the specified table, including all indexes and TOAST data|

#### Sample Usage [](hypertable_size-examples)
Get size information for a hypertable.
```sql
SELECT hypertable_size('devices') ;

 hypertable_size
-----------------
           73728
```
---

> ## hypertable_index_size() :community_function: [](hypertable_index_size)

Reviewer: Remove :community_function:

total | 48 kB

```
---

> ## hypertable_detailed_size() :community_function: [](hypertable_detailed_size)

Reviewer: Isn't this in Apache-2 codebase, not TSL?  If so, it shoudn't be marked :community_function:

(Same comment throughout)

Author: @mfreed The size functions and some of the new utility functions make calls to tsl code in case the table is a distributed table. But this is not the case for regular hypertables.
So it would work for regular hypertable with Apache. But the function could call tsl code in some circumstances. So is this Apache or Community?

Reviewer 2: you can make the function apache and still have the call to tsl code

Reviewer 2: i mean have the call be conditional and run tsl code only in community edition we have a couple of places that do that already like telemetry that gets additional information

Author: This calls tsl code for distributed hypertables. Since we cannot have distributed hypertables in apache, is this good enough?
```
        CASE WHEN is_distributed THEN
            RETURN QUERY SELECT * FROM _timescaledb_internal.hypertable_remote_size(schema_name, table_name);
        ELSE
            RETURN QUERY SELECT *, NULL::name FROM _timescaledb_internal.hypertable_local_size(schema_name, table_name);
        END CASE;
```
The remote_size function invokes the tsl code.

Reviewer 2: yes i think this is fine

Reviewer: I believe this should then be updated to remove the :community_function: label?

Author: removed community_function label.

> Get size of hypertable like `pg_relation_size(hypertable)`.
> If this is a distributed hypertable, the function returns size
> information per node. All sizes are reported in bytes.

Reviewer: ```suggestion
Get size of hypertable like `pg_relation_size(hypertable)`, returning 
size information for the table itself, any indexes on the table, any 
toast tables, and the total size of all. All sizes are reported in bytes.

If this is a distributed hypertable, the function returns size
information as a separate row per node. 
```

#### Required Arguments [](hypertable_detailed_size-required-arguments)

|Name|Description|
|---|---|

> | `main_table` | Identifier of hypertable to get relation size for.|

Reviewer: Why isn't the name of the argument "hypertable"?

Also, in the chunks_* docs, it describes this as:  "Name of the hypertable", but here as "Identifier of hypertable to get relation size for."

Should this just be "Name of the hypertable"?

It can be confusing if we describe things differently even though we mean the same thing.  (Also, "identifier" here is slightly odd choice of language.)

(Same question for other functions here.)

#### Returns [](hypertable_detailed_size-returns)
|Column|Description|
|---|---|
|table_bytes|Disk space used by main_table (like pg_relation_size(main_table))|
|index_bytes|Disk space used by indexes|
|toast_bytes|Disk space of toast tables|
|total_bytes|Total disk space used by the specified table, including all indexes and TOAST data|
|node_name| Node for which size is reported, applicable only to distributed hypertables|

#### Sample Usage [](hypertable_detailed_size-examples)
Get size information for a hypertable.
```sql
-- disttable is a distributed hypertable --
SELECT * FROM hypertable_detailed_size('disttable') ORDER BY node_name;

 table_bytes | index_bytes | toast_bytes | total_bytes |  node_name
-------------+-------------+-------------+-------------+-------------
       16384 |       32768 |           0 |       49152 | data_node_1
        8192 |       16384 |           0 |       24576 | data_node_2

```
---

## chunks_detailed_size()  :community_function: [](chunks_detailed_size)

> Get per chunk size information of hypertable.
> This function will return multiple rows of output for a single
> chunk, if this is a distributed hypertable. Meta data for a chunk can be
> viewed using `timescaledb_information.chunks` view.

Reviewer: I like re-using language across docs for consistency whenever possible.

```suggestion
Get size information about the chunks belonging to a hypertable, returning 
size information for each chunk table itself, any indexes on the chunk, any 
toast tables, and the total size associated with the chunk. All sizes are reported in bytes.

If this is a distributed hypertable, the function returns [size
information as a separate row per node - SEE COMMENT].

Additional metadata associated with a chunk can be accessed 
via the `timescaledb_information.chunks` view.
```

I don't understand this:  
```
This function will return multiple rows of output for a single
chunk, if this is a distributed hypertable. 
```
Is this referring to the case when replication_factor > 1?  Otherwise, shouldn't there only be a single copy of a chunk, on a specific data node?

#### Required Arguments [](chunks_detailed_size-required-arguments)

|Name|Description|
|---|---|

> | `main_table` | Name of the hypertable|

Reviewer: Given the description, this only works for hypertables, and not regular tables, right?

Why isn't the name of the argument "hypertable"?

#### Returns [](chunks_detailed_size-returns)
|Column|Description|
|---|---|
|chunk_schema| Schema name of the chunk |
|chunk_name| Name of the chunk|
|table_bytes|Disk space used by the chunk table|
|index_bytes|Disk space used by indexes|
|toast_bytes|Disk space of toast tables|
|total_bytes|Total disk space used by the chunk, including all indexes and TOAST data|
|node_name| Node for which size is reported, applicable only to distributed hypertables|

#### Sample Usage [](chunks_detailed_size-examples)
```sql

> SELECT * FROM chunks_detailed_size('dist_table')
> ORDER BY chunk_name, node_name;
>      chunk_schema      |      chunk_name       | table_bytes | index_bytes | toast_bytes | total_bytes |       node_name

Reviewer: ```suggestion
SELECT * FROM chunks_detailed_size('dist_table')
  ORDER BY chunk_name, node_name;
  
     chunk_schema      |      chunk_name       | table_bytes | index_bytes | toast_bytes | total_bytes |       node_name
```

-----------------------+-----------------------+-------------+-------------+-------------+-------------+-----------------------
 _timescaledb_internal | _dist_hyper_1_1_chunk |        8192 |       32768 |           0 |       40960 | db_node1

>  _timescaledb_internal | _dist_hyper_1_1_chunk |        8192 |       32768 |           0 |       40960 | db_node2

Reviewer: See my question above about why a chunk appears on multiple data nodes?  Given that (at least for now) we don't recommend replication_factor > 1, if that's the only case for it, I would simplify this example to show replication_factor = 1.  We can extend these docs once we launch more complete support for replication.

Author: Modified. I wasn't aware that we don't recommend this usage.

 _timescaledb_internal | _dist_hyper_1_2_chunk |        8192 |       32768 |           0 |       40960 | db_node2
 _timescaledb_internal | _dist_hyper_1_2_chunk |        8192 |       32768 |           0 |       40960 | db_node3
```
---

## hypertable_size() [](hypertable_size)

Get the total disk space used bya  table, including all indexes and TOAST tables.
The size is reported in bytes.
This is equivalent to computing the sum of `total_bytes` column from the output of 

> `hypertable_detailed_size` function.

Reviewer: Please align language with other docs.

#### Required Arguments [](hypertable_size-required-arguments)

|Name|Description|
|---|---|
| `main_table` | Name of the hypertable|

#### Returns [](hypertable_size-returns)

> Total disk space used by the specified table, including all indexes and TOAST data|

Reviewer: ```suggestion
Total disk space used by the specified table, including all indexes and TOAST data
```

#### Sample Usage [](hypertable_size-examples)
Get size information for a hypertable.
```sql

> SELECT hypertable_size('disttable') ;
>  hypertable_size

Reviewer: ```suggestion
SELECT hypertable_size('disttable') ;

 hypertable_size
```

-----------------
           73728
```
---

> ## hypertable_index_size() [](indexes_relation_size)

Reviewer: Shouldn't the anchor tag be "hypertable_index_size" ?

Get size of an index on a hypertable.

#### Required Arguments [](indexes_relation_size-required-arguments)

|Name|Description|
|---|---|

> | `main_table` | Name of hypertable|

Reviewer: Isn't the required arg "index" not "hypertable" name?

#### Returns [](indexes_relation_size-returns)
Returns disk space used by the index. 

#### Sample Usage [](indexes_relation_size-examples)

Get size of a specific index on a hypertable.

```sql
\d conditions_table
                     Table "public.test_table"
 Column |           Type           | Collation | Nullable | Default 
--------+--------------------------+-----------+----------+---------
 time   | timestamp with time zone |           | not null | 
 device | integer                  |           |          | 
 volume | integer                  |           |          | 
Indexes:
    "second_index" btree ("time")
    "test_table_time_idx" btree ("time" DESC)
    "third_index" btree ("time")

select pg_size_pretty(hypertable_index_size('second_index'));
 pg_size_pretty 
----------------

>  160 kB

Reviewer: Maybe should example first without pg_size_pretty?

(Also, for consistency, I think we are capitalizing "SELECT")

and then inspect `dump_file.txt` before sending it together with a bug report or support question.

[Slack]: https://slack-login.timescale.com

> [chunk detailed size]: #chunk_detialed_size

Reviewer: ```suggestion
[chunk detailed size]: #chunk_detailed_size
```

and then inspect `dump_file.txt` before sending it together with a bug report or support question.

[Slack]: https://slack-login.timescale.com

> [chunk detailed size]: #chunk_detailed_size

Reviewer: I don't /think/ this is used anywhere?  But regardless not incorrect.
